### ISSUE TITLE
Coral Problem

### ISSUE THREAD

**Issue by SherryJingxr** [wzell/mann#13]
I would like to thank you so much for your great work！
I used my own data set and modified the structure of the network. I wanted to use the coral method and ran these codes in sentiment_analysis.py file. 
![image](https://user-images.githubusercontent.com/100208141/179971136-b1450c8b-d0a5-49ce-a071-e8375609ea9c.png)

Finally, I found that the effect before and after the coral was the same. I don’t understand the code mechanism in the coral. I don't think the coral mechanism is working, but don't know what's wrong
![image](https://user-images.githubusercontent.com/100208141/179971359-47d9ca1b-15af-487d-bd75-21bfcfabc1fc.png)
I try to print a sentence in the coral function, obviously this sentence is not executed at the end.
![image](https://user-images.githubusercontent.com/100208141/179971573-bbcd678f-0aec-4e03-9f58-f811bb13f124.png)

**Comment by wzell** [comment]
Have you tried more than one epoch, different regularization parameters, re-initialization of environment?

**Comment by SherryJingxr** [comment]
> regularization

I'm very sorry, is the epoch adjusted by the MAX_N_EPOCH parameter in the first picture? And how to realise re-initialization of environment?

**Comment by SherryJingxr** [comment]
> Have you tried more than one epoch, different regularization parameters, re-initialization of environment?

I'm very sorry, is the epoch adjusted by the MAX_N_EPOCH parameter in the first picture? And how to realise re-initialization of environment?

**Comment by wzell** [comment]
(a) Yes, it is. (b) Re-initialization of environment depends on your versions of software libraries, e.g. use https://www.tensorflow.org/api_docs/python/tf/keras/backend/clear_session for newer version of keras. The point is to make sure that the second model does not depend on the first one, but it is a new (keras) model.

**Comment by SherryJingxr** [comment]
> (a) Yes, it is. (b) Re-initialization of environment depends on your versions of software libraries, e.g. use https://www.tensorflow.org/api_docs/python/tf/keras/backend/clear_session for newer version of keras. The point is to make sure that the second model does not depend on the first one, but it is a new (keras) model.

I tried the methods you mentioned, changed the epoch to 1000 times, and re-initialized the environment, but the final result is still the same, the statement that I tried to print in the third picture is still not printed, this is normal?

**Comment by wzell** [comment]
Yes, not-in-time-executed numpy code in theano/tensorflow functions executed by old keras version, can happen.
Did you try to increase the regularization parameter to 1000? To see if your own data (scaling) needs different parametrization..
Another question: does the simple example https://github.com/wzell/mann/blob/master/artificial_example.py work?

**Comment by SherryJingxr** [comment]
First of all, thank you very much for your patient help!
I just tried to run artificial_example.py with keras2.0.
First of all, it reported an error in the neural_network function, it should be a version problem because there is no merge function in 2.0 version, I replaced 
`dense_s_t = merge([encoded_s,encoded_t], mode='concat', concat_axis=1)`
with
`dense_s_t = Concatenate()([encoded_s, encoded_t])`
Then report:
![image](https://user-images.githubusercontent.com/100208141/180161059-eb03c212-9554-447d-8af4-d752261e45f2.png)
where 12261 is the number of source data, 17807 is the number of target data.

**Comment by wzell** [comment]
Nice spot with the version problem :-).
This seems to be because the zeros in line 215 are obtained from the shape of source sample. I suggest to exclude some target data so that the batches have the same size, and, then to try again. Later, if you find the method useful, you can solve this issue that arises from your data.

**Comment by SherryJingxr** [comment]
> Nice spot with the version problem :-). This seems to be because the zeros in line 215 are obtained from the shape of source sample. I suggest to exclude some target data so that the batches have the same size, and, then to try again. Later, if you find the method useful, you can solve this issue that arises from your data.

It's a good news that it worked! 
I think it should be the last question that the performance improvement on the target data is not very obvious,  I found val_activation_accuray is nearly 80%, but the val_activation_1_accuracy is only 20%, what does it mean? Is it overfitting?

**Comment by SherryJingxr** [comment]
> Nice spot with the version problem :-). This seems to be because the zeros in line 215 are obtained from the shape of source sample. I suggest to exclude some target data so that the batches have the same size, and, then to try again. Later, if you find the method useful, you can solve this issue that arises from your data.

It's a good news that it worked!
I think it should be the last question that the performance improvement on the target data is not very obvious, I found val_activation_accuray was nearly 80%, but the val_activation_1_accuracy was only 20%. What does it mean? Is it overfitting?

**Comment by SherryJingxr** [comment]
> Nice spot with the version problem :-). This seems to be because the zeros in line 215 are obtained from the shape of source sample. I suggest to exclude some target data so that the batches have the same size, and, then to try again. Later, if you find the method useful, you can solve this issue that arises from your data.

Sorry to bother you again!
I would like to know how to evaluate the accuracy on the target domain at the end? By running the accuracy function?
![image](https://user-images.githubusercontent.com/100208141/181229746-8c41d802-460b-464e-81ff-e83881f36962.png)

I ran the accuracy function and got a result, about 0.5, and the accuracy of the source domain also drops
But I passed the model through the model.evaluate function that comes with keras 2.0, and the accuracy I got was extremely low, only 0.19. Did I make a mistake?

**Comment by wzell** [comment]
Unfortunately, it is very hard for me to help without knowing anything about your data. I just can remark that the regularization weight has ussually a very high influence on the accuracy.

Since we were able to figure out the problem of the CORAL-loss, I allow myself to close this issue.
